src/statement.ts

```diff
@@ -1,416 +1,467 @@
+  columnName(index: number): string {
+    const cached = this.#colNameCache[index];
+    if (cached !== undefined) return cached;
+    return (this.#colNameCache[index] = readCstr(
+      readCstr(sqlite3_column_name(this.#handle, index)),
+    ));
   }
 
-  constructor(db: Database, handle: sqlite3_stmt) {
-    this.#db = db;
-    this.#handle = handle;
+  #cachedColCount?: number;
+
+  /** Column count in current row. */
+  get columnCount(): number {
+    if (this.#cachedColCount !== undefined) return this.#cachedColCount;
+    return (this.#cachedColCount = sqlite3_column_count(this.#handle));
   }
 
-  /** Binding parameter count in the prepared statement. */
-  get bindParameterCount(): number {
-    return sqlite3_bind_parameter_count(this.#handle);
+  #begin(): void {
+    sqlite3_reset(this.#handle);
+    sqlite3_clear_bindings(this.#handle);
+    this.#cachedColCount = undefined;
+    this.#colNameCache = {};
   }
 
-  /** Get name of a binding parameter by its index. */
-  bindParameterName(index: number): string {
-    return sqlite3_bind_parameter_name(this.#handle, index);
+  #getColumn(handle: number, i: number): any {
+    const ty = sqlite3_column_type(handle, i);
+    if (ty === SQLITE_INTEGER) return sqlite3_column_int(handle, i);
+    switch (ty) {
+      case SQLITE_TEXT: {
+        const ptr = sqlite3_column_text(handle, i);
+        if (ptr === 0) return null;
+        return readCstr(ptr);
+      }
+
+      case SQLITE_INTEGER: {
+        return sqlite3_column_int(handle, i);
+      }
+
+      case SQLITE_FLOAT: {
+        return sqlite3_column_double(handle, i);
+      }
+
+      case SQLITE_BLOB: {
+        const ptr = sqlite3_column_blob(handle, i);
+        const bytes = sqlite3_column_bytes(handle, i);
+        return new Uint8Array(
+          new Deno.UnsafePointerView(BigInt(ptr)).getArrayBuffer(bytes)
+            .slice(0),
+        );
+      }
+
+      default: {
+        return null;
+      }
+    }
   }
 
-  /** Get index of a binding parameter by its name. */
-  bindParameterIndex(name: string): number {
-    const index = sqlite3_bind_parameter_index(this.#handle, name);
-    if (index === 0) {
-      throw new Error(`Couldn't find index for '${name}'`);
+  #runNoArgs(): undefined {
+    this.#begin();
+    const status = sqlite3_step(this.#handle);
+    if (status === SQLITE3_ROW || status === SQLITE3_DONE) {
+      return undefined;
+    } else {
+      unwrap(status, this.db.unsafeHandle);
     }
-    return index;
   }
 
-  /**
-   * We need to store references to any type that involves passing pointers
-   * to avoid V8's GC deallocating them before the statement is finalized.
-   *
-   * In SQLite C API, there is a callback that we can pass for such types
-   * to deallocate only when they're not in use. But this is not possible
-   * using Deno FFI. So we will just store references to them until `finalize`
-   * is called.
-   */
-  #bufferRefs = new Set<Uint8Array>();
```

Comment:
Do we still need this...

Reply:
nope synchronous calls...no need to get buffer references